libcloud/common/openstack_identity.py

```
@@ -1350,6 +1450,93 @@ def disable_user(self, user):
         user = self._to_user(data=response.object['user'])
         return user
 
+    def _fetch_auth_context_from_cache(self):
+        context = super()._fetch_auth_context_from_cache()
+        if context is None:
+            return None
+
+        # Since v3 only caches the token and expiration, fetch the
+        # service catalog and other bits of the authentication context
+        # from Keystone.
+        try:
+            self._fetch_auth_token()
+        except InvalidCredsError:  # Unauthorized
+            self.clear_cached_auth_context()
+            return None
+
+        # Local auth context variables set in _fetch_auth_token
+        return context
+
+    def _parse_token_response(self, response, cache_it=False,
+                              raise_ambiguous_version_error=True):
+        """
+        Parse a response from /v3/auth/tokens.
+
+        :param cache_it: Should we cache the authentication context?
+        :type cache_it: ``bool``
+
+        :param raise_ambiguous_version_error: Should an ambiguous version
+            error be raised on a 300 response?
+        :type raise_ambiguous_version_error: ``bool``
+        """
+        if response.status == httplib.UNAUTHORIZED:
+            raise InvalidCredsError()
+        elif response.status in [httplib.OK, httplib.CREATED]:
+            headers = response.headers
+
+            try:
+                body = json.loads(response.body)
+            except Exception as e:
+                raise MalformedResponseError('Failed to parse JSON', e)
+
+            try:
+                roles = self._to_roles(body['token']['roles'])
+            except Exception:
```

Could catch a more specific Exception here? Otherwise, unexpected Exceptions are just swallowed.

---

Same copy/paste.